### PR TITLE
Feat/fp

### LINKED ISSUE #4: Fp type
The field that corresponds to the coordinates of the points on the elliptic curve group should have a go wrapper type that exposes all of the needed methods.

### PR DESCRIPTION
**Pull request by ross-pure**
This PR resolves #4.

### CODE REVIEW

File: fp.go [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+// AddAssignUnsafe computes the addition of the receiver and the argument and
+// stores the result in the receiver.
+//
+// Unsafe: If this function receives nil arguments, the behaviour is
+// implementation dependent, because the definition of the NULL pointer in c is
+// implementation dependent. If the NULL pointer and the go nil pointer are the
+// same, then the function will panic.
+func (x *Fp) AddAssignUnsafe(a *Fp) {
+	C.secp256k1_fe_add((*C.secp256k1_fe)(unsafe.Pointer(x)), (*C.secp256k1_fe)(unsafe.Pointer(a)))
```
Comment (loongy) [inline fp.go]:
```suggestion
	C.secp256k1_fe_add((*C.secp256k1_fe)(unsafe.Pointer(&x.inner[0])), (*C.secp256k1_fe)(unsafe.Pointer(&a.inner[0])))
```

File: fp.go [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+// NOTE: This function performs a copy of one of the arguments. To avoid this
+// copy, use AddAssign.
+//
+// Unsafe: If this function receives nil arguments, the behaviour is
+// implementation dependent, because the definition of the NULL pointer in c is
+// implementation dependent. If the NULL pointer and the go nil pointer are the
+// same, then the function will panic.
+func (x *Fp) AddUnsafe(a, b *Fp) {
+	aCopy := *a
+	C.secp256k1_fe_add((*C.secp256k1_fe)(unsafe.Pointer(&aCopy)), (*C.secp256k1_fe)(unsafe.Pointer(b)))
```
Comment (loongy) [inline fp.go]:
```suggestion
	C.secp256k1_fe_add((*C.secp256k1_fe)(unsafe.Pointer(&aCopy.inner[0])), (*C.secp256k1_fe)(unsafe.Pointer(&b.inner[0])))
```

File: fp.go [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+// form.
+//
+// Panics: If the byte slice has length less than 32, this function will panic.
+func (x Fp) PutB32(dst []byte) {
+	if len(dst) < 32 {
+		panic(fmt.Sprintf("invalid slice length: length needs to be at least 32, got %v", len(dst)))
+	}
+
+	// NOTE: This function assumes that the representation is normalised.
+	C.secp256k1_fe_get_b32((*C.uchar)(&dst[0]), (*C.secp256k1_fe)(unsafe.Pointer(&x)))
```
Comment (loongy) [inline fp.go]:
```suggestion
	C.secp256k1_fe_get_b32((*C.uchar)(&dst[0]), (*C.secp256k1_fe)(unsafe.Pointer(&x.inner[0])))
```

File: fp.go [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+// function will return true if the bytes represented a number greater than or
+// equal to P, and false otherwise.
+//
+// Panics: If the byte slice has length less than 32, this function will panic.
+func (x *Fp) SetB32(bs []byte) bool {
+	if len(bs) < 32 {
+		panic(fmt.Sprintf("invalid slice length: length needs to be at least 32, got %v", len(bs)))
+	}
+
+	greater := int(C.secp256k1_fe_set_b32((*C.secp256k1_fe)(unsafe.Pointer(x)), (*C.uchar)(&bs[0]))) == 0
```
Comment (loongy) [inline fp.go]:
```suggestion
	greater := int(C.secp256k1_fe_set_b32((*C.secp256k1_fe)(unsafe.Pointer(&x.inner[0])), (*C.uchar)(&bs[0]))) == 0
```

Comment (loongy) [reply fp.go]: There is a stronger guarantee that a pointer to the first element of the contiguous array will be forwards compatible than a pointer to the struct itself (which has no guarantee in Go to not change in future versions, w.r.t. `unsafe.Pointer`).

File: fp.go [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+// the result in the receiver.
+//
+// Unsafe: If this function receives nil arguments, the behaviour is
+// implementation dependent, because the definition of the NULL pointer in c is
+// implementation dependent. If the NULL pointer and the go nil pointer are the
+// same, then the function will panic.
+func (x *Fp) NegateUnsafe(a *Fp) {
+	// NOTE: The final argument is set to 0 because it is assumed that the
+	// representation is normalized.
+	C.secp256k1_fe_negate((*C.secp256k1_fe)(unsafe.Pointer(x)), (*C.secp256k1_fe)(unsafe.Pointer(a)), 0)
```
Comment (loongy) [inline fp.go]:
```suggestion
	C.secp256k1_fe_negate((*C.secp256k1_fe)(unsafe.Pointer(&x.inner[0])), (*C.secp256k1_fe)(unsafe.Pointer(&a.inner[0])), 0)
```

File: fp.go [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+// implementation dependent, because the definition of the NULL pointer in c is
+// implementation dependent. If the NULL pointer and the go nil pointer are the
+// same, then the function will panic.
+func (x *Fp) MulNoAliaseUnsafe(a, b *Fp) {
+	// NOTE: The c function defines the pointer to b as restrict, which means
+	// that it must not be aliased by x or a.
+	C.secp256k1_fe_mul(
+		(*C.secp256k1_fe)(unsafe.Pointer(x)),
+		(*C.secp256k1_fe)(unsafe.Pointer(a)),
+		(*C.secp256k1_fe)(unsafe.Pointer(b)),
```
Comment (loongy) [inline fp.go]:
```suggestion
		(*C.secp256k1_fe)(unsafe.Pointer(&x.inner[0])),
		(*C.secp256k1_fe)(unsafe.Pointer(&a.inner[0])),
		(*C.secp256k1_fe)(unsafe.Pointer(&b.inner[0])),
```

File: fp.go [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+// same, then the function will panic.
+func (x *Fp) MulUnsafe(a, b *Fp) {
+	// NOTE: The c function defines the pointer to b as restrict, which means
+	// that it must not be aliased by x or a. We therefore make a new copy of b
+	// to ensure that there is no aliasing.
+	bCopy := *b
+	C.secp256k1_fe_mul(
+		(*C.secp256k1_fe)(unsafe.Pointer(x)),
+		(*C.secp256k1_fe)(unsafe.Pointer(a)),
+		(*C.secp256k1_fe)(unsafe.Pointer(&bCopy)),
```
Comment (loongy) [inline fp.go]:
```suggestion
		(*C.secp256k1_fe)(unsafe.Pointer(&x.inner[0])),
		(*C.secp256k1_fe)(unsafe.Pointer(&a.inner[0])),
		(*C.secp256k1_fe)(unsafe.Pointer(&bCopy.inner[0])),
```

File: fp.go [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+
+// SqrUnsafe computes the square of the field element and stores the result in
+// the receiver.
+//
+// Unsafe: If this function receives nil arguments, the behaviour is
+// implementation dependent, because the definition of the NULL pointer in c is
+// implementation dependent. If the NULL pointer and the go nil pointer are the
+// same, then the function will panic.
+func (x *Fp) SqrUnsafe(a *Fp) {
+	C.secp256k1_fe_sqr((*C.secp256k1_fe)(unsafe.Pointer(x)), (*C.secp256k1_fe)(unsafe.Pointer(a)))
```
Comment (loongy) [inline fp.go]:
```suggestion
	C.secp256k1_fe_sqr((*C.secp256k1_fe)(unsafe.Pointer(&x.inner[0])), (*C.secp256k1_fe)(unsafe.Pointer(&a.inner[0])))
```

File: fp.go [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+// stores the result in the receiver.
+//
+// Unsafe: If this function receives nil arguments, the behaviour is
+// implementation dependent, because the definition of the NULL pointer in c is
+// implementation dependent. If the NULL pointer and the go nil pointer are the
+// same, then the function will panic.
+func (x *Fp) InvUnsafe(a *Fp) {
+	// We use the potentially faster but not constant time version of the
+	// inverse function.
+	C.secp256k1_fe_inv_var((*C.secp256k1_fe)(unsafe.Pointer(x)), (*C.secp256k1_fe)(unsafe.Pointer(a)))
```
Comment (loongy) [inline fp.go]:
```suggestion
	C.secp256k1_fe_inv_var((*C.secp256k1_fe)(unsafe.Pointer(&x.inner[0])), (*C.secp256k1_fe)(unsafe.Pointer(&a.inner[0])))
```

File: fp.go [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+func (x *Fp) Eq(other *Fp) bool {
+	return ((x.inner[0] ^ other.inner[0]) |
+		(x.inner[1] ^ other.inner[1]) |
+		(x.inner[2] ^ other.inner[2]) |
+		(x.inner[3] ^ other.inner[3]) |
+		(x.inner[4] ^ other.inner[4])) == 0
+}
+
+func (x *Fp) normalize() {
+	C.secp256k1_fe_normalize_var((*C.secp256k1_fe)(unsafe.Pointer(x)))
```
Comment (loongy) [inline fp.go]:
```suggestion
	C.secp256k1_fe_normalize_var((*C.secp256k1_fe)(unsafe.Pointer(&x.inner[0])))
```

File: fp.go [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+}
+
+// IsOne returns true if the field element is zero and false otherwise.
+func (x *Fp) IsOne() bool {
+	return (x.inner[0] == 1) && ((x.inner[1] | x.inner[2] | x.inner[3] | x.inner[4]) == 0)
+}
+
+// IsEven returns true if the field element is even and false otherwise.
+func (x *Fp) IsEven() bool {
+	return !((x.inner[0] & 1) == 1)
```
Comment (loongy) [inline fp.go:397]: Confirming, this is meant to be `x.inner[0]` and not `x.inner[4]`?

**Review by loongy — APPROVED** [review]
@ross-pure LGTM

File: fp.go (hunk 11 above)
Comment (ross-pure) [reply fp.go:397]: Yep, `x.inner[0]` is the least significant limb.

**Pull request merged**